plugins/basic/nexus-repository-docker/src/main/java/org/sonatype/nexus/repository/docker/internal/DockerFacetImpl.java

```diff
@@ -0,0 +1,669 @@
+import static org.sonatype.nexus.repository.storage.StorageFacet.P_VERSION;
+
+/**
+ * {@link DockerFacet} implementation.
+ *
+ * @since 3.0
+ */
+@Named
+public class DockerFacetImpl
+    extends FacetSupport
+    implements DockerFacet
+{
+
+  private final static List<HashAlgorithm> hashAlgorithms = Lists.newArrayList(SHA256);
+
+  private final static ObjectMapper mapper = new ObjectMapper();
+
+  private final static Pattern validLayerId = Pattern.compile("^([a-f0-9]{16}|[a-f0-9]{64})$");
+
+  private final static Pattern validTag = Pattern.compile("^[\\w][\\w.-]{0,127}$");
+
+
+  @Override
+  @Transactional(retryOn = IllegalStateException.class)
+  public Payload getImagesIndex(final String name) throws IOException {
+    StorageTx tx = UnitOfWork.currentTransaction();
+
+    String imagesIndexAssetName = name + "/images";
+    Bucket bucket = tx.getBucket();
+    // check that we have images index asset
+    Asset asset = findAsset(tx, bucket, imagesIndexAssetName);
+    if (asset == null) {
+      throw new V1Exception("images not found", HttpStatus.NOT_FOUND);
+    }
+
+    return marshall(asset, tx.requireBlob(asset.requireBlobRef()));
+  }
+
+  @Override
+  public void putImagesIndex(final String name, final Payload content) throws IOException {
+    try (TempStreamSupplier streamSupplier = new TempStreamSupplier(content.openInputStream())) {
+      readImagesIndex(streamSupplier);
+      doPutImagesIndex(name, streamSupplier, content);
+    }
+  }
+
+  @Transactional(retryOn = {ONeedRetryException.class, ORecordDuplicatedException.class})
+  protected void doPutImagesIndex(final String name,
+                                  final Supplier<InputStream> streamSupplier,
+                                  final Payload payload)
+      throws IOException
+  {
+    StorageTx tx = UnitOfWork.currentTransaction();
+
+    InputStream indexStream;
+
+    String imagesIndexAssetName = name + "/images";
+
+    Bucket bucket = tx.getBucket();
+    Asset imagesIndexAsset = findAsset(tx, bucket, imagesIndexAssetName);
+    if (imagesIndexAsset == null) {
+      imagesIndexAsset = tx.createAsset(bucket, getRepository().getFormat());
+      imagesIndexAsset.name(imagesIndexAssetName);
+      imagesIndexAsset.formatAttributes().set(P_PATH, "/v1/repositories/" + imagesIndexAssetName);
+
+      indexStream = streamSupplier.get();
+    }
+    else {
+      // merge existing index with the new one
+      List<Map<String, Object>> existingIndex = readImagesIndex(blobStreamSupplier(tx, imagesIndexAsset));
+      List<Map<String, Object>> newIndex = readImagesIndex(streamSupplier);
+      List<Map<String, Object>> mergedIndex = new ArrayList<>(existingIndex);
+      for (Map<String, Object> newImage : newIndex) {
+        boolean found = false;
+        for (Map<String, Object> existingImage : existingIndex) {
+          if (Objects.equals(existingImage.get("id"), newImage.get("id")) &&
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS134)